src/crh/poseidon/mod.rs

```diff
@@ -31,191 +49,210 @@ pub trait PoseidonRoundParams<F: PrimeField>: Default + Clone {
 
 /// The Poseidon permutation.
 #[derive(Default, Clone)]
-pub struct Poseidon<F, P> {
-    pub params: P,
+pub struct PoseidonParameters<F> {
     /// The round key constants
     pub round_keys: Vec<F>,
     /// The MDS matrix to apply in the mix layer.
     pub mds_matrix: Vec<Vec<F>>,
 }
 
-impl<F: PrimeField, P: PoseidonRoundParams<F>> Poseidon<F, P> {
-    fn permute(&self, input: &[F]) -> Vec<F> {
-        let width = P::WIDTH;
-        assert_eq!(input.len(), width);
+impl<F: PrimeField> PoseidonParameters<F> {
+    pub fn new(round_keys: Vec<F>, mds_matrix: Vec<Vec<F>>) -> Self {
+        Self {
+            round_keys,
+            mds_matrix,
+        }
+    }
 
-        let full_rounds_beginning = P::FULL_ROUNDS_BEGINNING;
-        let partial_rounds = P::PARTIAL_ROUNDS;
-        let full_rounds_end = P::FULL_ROUNDS_END;
+    pub fn generate<R: Rng>(rng: &mut R) -> Self {
+        Self {
+            round_keys: Self::create_round_keys(rng),
+            mds_matrix: Self::create_mds(rng),
+        }
+    }
+    pub fn create_mds<R: Rng>(_rng: &mut R) -> Vec<Vec<F>> {
+        let mds_matrix = Vec::new();
+        mds_matrix
+    }
 
-        let mut current_state = input.to_vec();
-        let mut current_state_temp = vec![F::zero().clone(); width];
+    pub fn create_round_keys<R: Rng>(_rng: &mut R) -> Vec<F> {
+        let round_keys = Vec::new();
+        round_keys
+    }
+}
+
+pub struct CRH<F: PrimeField, P: Rounds> {
+    field: PhantomData<F>,
+    rounds: PhantomData<P>,
+}
+
+impl<F: PrimeField, P: Rounds> CRH<F, P> {
+    fn permute(params: &PoseidonParameters<F>, mut state: Vec<F>) -> Result<Vec<F>, PoseidonError> {
+        let width = P::WIDTH;
 
         let mut round_keys_offset = 0;
 
         // full Sbox rounds
-        for _ in 0..full_rounds_beginning {
+        for _ in 0..(P::FULL_ROUNDS / 2) {
             // Sbox layer
             for i in 0..width {
-                current_state[i] += self.round_keys[round_keys_offset];
-                current_state[i] = P::SBOX.apply_sbox(current_state[i]);
+                state[i] += params.round_keys[round_keys_offset];
+                state[i] = P::SBOX.apply_sbox(state[i])?;
                 round_keys_offset += 1;
             }
-
             // linear layer
-            for j in 0..width {
-                for i in 0..width {
-                    current_state_temp[i] += current_state[j] * self.mds_matrix[i][j];
-                }
-            }
-
-            // Output of this round becomes input to next round
-            for i in 0..width {
-                current_state[i] = current_state_temp[i];
-                current_state_temp[i] = F::zero();
-            }
+            state = Self::apply_linear_layer(&state, &params.mds_matrix);
```

Comment:
Maybe this should mutate `state` in place? To avoid unnecessary allocations?

Reply:
Hmm I see that we can't easily do that...

Maybe we should replace `Vec` with `SmallVec` here, so that we don't incur unnecessary allocations here.

Reply:
That would require me to use const generic for the argument `SmallVec<[F; P::WIDTH]>` for which I have to add these:
```
#![allow(incomplete_features)]
#![feature(const_generics)]
#![feature(const_evaluatable_checked)]
```

Also, why we are at it, why not just use array `[F; P::WIDTH]`?

Reply:
We can also have parameter sizes bound to const generic, something like:
```rust
pub struct PoseidonParameters<F, P: Rounds> {
    pub round_keys: [F; P::PARTIAL_ROUNDS * P::WIDTH + P::FULL_ROUNDS * P::WIDTH],
    pub mds_matrix: [[F; P::WIDTH]; P::WIDTH],
}
```